Convert piles from concatenated to stacked-and-padded, and revive existing tests

axch opened the pull request.

axch committed:
Redefine the pile representation from concatenated to stacked-and-padded.

The advantage (already being realized) is that the batching rules
become much simpler: we just batch along the stacked axis as always,
and when a reduction is about to occur, also mask out the padding
elements, replacing them with the identity element of the reduction.

This commit

- Changes the intended representation of data for piles and the
  corresponding BatchTracers.
- Re-defines ConcatAxis as RaggedAxis to represent the metadata.
- Updates `defreducer` to require the identity function (in case
  masking is needed), and supplies it everywhere.
- Flushes batching.segment_sum, as it is dead code now.
- Deletes unpack_concat_axes and reassemble_concat_axes, because they
  are irrelevant to the padded representation.

mattjj committed:
Improve printing of bints and piles, and allow bints in convert_element_type.

axch committed:
Fix flake8.

axch committed:
Force the mask computation into 'int32' more consistently

so as to make tests pass with JAX_ENABLE_X64=1.

mattjj commented on jax/_src/core.py:
```diff
@@ -1729,6 +1732,8 @@ def __eq__(self, other):
     if isinstance(other, DArray) and self._aval == other._aval:
       return self._data == other._data
     return False
+  def __len__(self):
```
Do we need this? If not, can we remove it?

axch replied:
Actually, we seem to -- the `segment_lengths` can be a `DArray`, and we query its `len`.

mattjj commented on jax/_src/interpreters/batching.py:
```diff
@@ -52,7 +51,8 @@ class PileTy:
   elt_ty: core.DShapedArray
   def __repr__(self) -> str:
-    return f'Var{id(self.binder)}:{self.length} => {self.elt_ty}'
+    # return f'Var{id(self.binder)}:{self.length} => {self.elt_ty}'
+    return f'i:{self.length} => {self.elt_ty}'
```
We probably want to revert this change, right?

axch replied:
Done.

mattjj commented on jax/_src/interpreters/batching.py:
```diff
@@ -61,7 +61,7 @@ class IndexedAxisSize:
   lengths: Union[Array, core.Var, Tracer]
   def __repr__(self) -> str:
-    return f'{str(self.lengths)}.Var{id(self.idx)}'
+    return f'{str(self.lengths)}.i'
```
Also revert this change?

axch replied:
Done.

mattjj commented on jax/_src/interpreters/batching.py:
```diff
@@ -376,12 +380,9 @@ def process_call(self, call_primitive, f, tracers, params):
-                            segment_lens)
-    vals_out = call_primitive.bind(f_, *segment_lens, *vals, **params)
-    vals_out, dims_out = reassemble_concat_axes(vals_out, dims_out())
+    f_ = _update_annotation(f_, f.in_type, axis_size, self.axis_name, dims, [])
```
This doesn't seem right in general; can you explain why we can always pass an empty list here, and skip the `unpack_concat_axis` / `reassemble_concat_axis` stuff? I assume we can't skip those things in general, but we don't have a test that's exercising this path. If so, we should figure out what needs to be done and leave a TODO (or else fix up this section).

axch replied:
Let's work this one out synchronously.  We don't need to call a function literally called `unpack_concat_axis` because there are no more concat axes; whatever translation this was trying to do should be simpler in the padded-and-stacked representation.  That said, I'm not sure what translation this was trying to do, and what test would exercise it.  I passed an empty list because that's what `unpack_concat_axis` would do on a non-ragged array, which got the existing tests passing.

axch replied:
In offline discussion, we determined: (i) this code seems to only relevant for getting axes to work under `jit`, which is why current tests pass without it, and (ii) we will revive an updated version in a future PR.

mattjj commented on jax/_src/interpreters/batching.py:
```diff
@@ -98,23 +98,28 @@ def _pile_unflatten(aval, x):
 @dataclasses.dataclass(frozen=True)
-class ConcatAxis:
-  axis: int
+class RaggedAxis:
```
Could you leave a TODO here, or in the PR message, explaining how we need to generalize `RaggedAxis` further, to handle piles like `i:(Fin 3) => f32[lens1.i, lens2.i]` (i.e. multiple ragged axes)?

axch replied:
Done.

axch committed:
Tidy up in response to first review.

mattjj reviewed (approved):
So sick, also the PRs to come.

axch committed:
Pass pytype.

The ignore is justified because `core.mapped_aval` preserves the
presence of a `.shape` attribute, which we already expect a couple of
lines up.
